src/components/KMultiselect/KMultiselect.vue

```diff
@@ -0,0 +1,911 @@
+    const handleItemSelect = (item: MultiselectItem) => {
+      const selectedItem = unfilteredItems.value.filter(anItem => anItem.key === item.key)[0]
+      // if clicked item is already selected
+      if (selectedItem.selected) {
+        selectedItems.value = selectedItems.value.filter(anItem => anItem.key !== item.key)
+        // remove item from visibility arrays
+        if (visibleSelectedItemsStaging.value.filter(anItem => anItem.key === item.key).length) {
+          visibleSelectedItemsStaging.value = visibleSelectedItemsStaging.value.filter(anItem => anItem.key !== item.key)
+        } else if (invisibleSelectedItemsStaging.value.filter(anItem => anItem.key === item.key).length) {
+          invisibleSelectedItemsStaging.value = invisibleSelectedItemsStaging.value.filter(anItem => anItem.key !== item.key)
+        }
+        // deselect item
+        selectedItem.selected = false
+        selectedItem.key = selectedItem?.key?.replace(/-selected/gi, '')
+
+        // if some items are hidden grab the first hidden one and add it into the visible array
+        if (invisibleSelectedItemsStaging.value.length) {
+          const item = invisibleSelectedItemsStaging.value.pop()
+          if (item) {
+            visibleSelectedItemsStaging.value.push(item)
+          }
+        }
+      } else { // newly selected item
+        selectedItem.selected = true
+        selectedItem.key = selectedItem?.key?.includes('-selected') ? selectedItem.key : `${selectedItem.key}-selected`
+        selectedItem.key += '-selected'
+        selectedItems.value.push(selectedItem)
+        visibleSelectedItemsStaging.value.push(selectedItem)
+      }
+
+      stageSelections()
+      const selectedVals = selectedItems.value.map(anItem => anItem.value)
+
+      emit('selected', selectedItems.value)
+      emit('update:modelValue', selectedVals)
+      // this 'input' event must be emitted for v-model binding to work properly
+      emit('input', selectedVals)
+      emit('change', item)
+    }
+
+    // sort dropdown items. Selected items displayed before unselected items
+    const sortItems = () => {
+      const selItems = filteredItems.value.filter((item: MultiselectItem) => item.selected)
+      const unselItems = filteredItems.value.filter((item: MultiselectItem) => !item.selected)
+
+      sortedItems.value = selItems.concat(unselItems)
+    }
+
+    const clearSelection = (): void => {
+      unfilteredItems.value.forEach(anItem => {
+        anItem.selected = false
+        anItem.key = anItem?.key?.replace(/-selected/gi, '')
+      })
+      selectedItems.value = []
+      visibleSelectedItemsStaging.value = []
+      invisibleSelectedItemsStaging.value = []
+      filterStr.value = ''
+      stageSelections()
+
+      emit('selected', [])
+      // this 'input' event must be emitted for v-model binding to work properly
+      emit('input', [])
+      emit('change', null)
+      emit('update:modelValue', [])
+      emit('query-change', '')
```

Comment:
nit: can these be put in the same order as the other emit block just for readability?